coderd/database/pubsub/pubsub_linux_test.go

```diff
@@ -294,3 +298,136 @@ func TestPubsub_Disconnect(t *testing.T) {
+		logger := slogtest.Make(t, nil).Leveled(slog.LevelDebug)
+		connectionURL, closePg, err := dbtestutil.Open()
+		require.NoError(t, err)
+		db, err := sql.Open("postgres", connectionURL)
+		require.NoError(t, err)
+		ps, err := pubsub.New(ctx, logger, db, connectionURL)
+		require.NoError(t, err)
+
+		return ps, func() {
+			_ = ps.Close()
+			_ = db.Close()
+			closePg()
+			cancel()
+		}
+	}
+
+	t.Run("MeasureLatency", func(t *testing.T) {
+		t.Parallel()
+
+		logger := slogtest.Make(t, nil).Leveled(slog.LevelDebug)
+		ps, done := newPubsub()
+		defer done()
+
+		ctx, cancel := context.WithTimeout(context.Background(), testutil.WaitShort)
+		defer cancel()
+
+		send, recv, err := pubsub.NewLatencyMeasurer(logger).Measure(ctx, ps)
+		require.NoError(t, err)
+		require.Greater(t, send, 0.0)
+		require.Greater(t, recv, 0.0)
+	})
+
+	t.Run("MeasureLatencyRecvTimeout", func(t *testing.T) {
+		t.Parallel()
+
+		logger := slogtest.Make(t, nil).Leveled(slog.LevelDebug)
+		ps, done := newPubsub()
+		defer done()
+
+		ctx, cancel := context.WithDeadline(context.Background(), time.Now().Add(-time.Second))
+		defer cancel()
+
+		send, recv, err := pubsub.NewLatencyMeasurer(logger).Measure(ctx, ps)
+		require.ErrorContains(t, err, context.DeadlineExceeded.Error())
+		require.Greater(t, send, 0.0)
+		require.EqualValues(t, recv, -1)
+	})
+
+	t.Run("MeasureLatencyNotifyRace", func(t *testing.T) {
+		t.Parallel()
+
+		var buf bytes.Buffer
+		logger := slogtest.Make(t, &slogtest.Options{IgnoreErrors: true}).Leveled(slog.LevelDebug)
+		logger = logger.AppendSinks(sloghuman.Sink(&buf))
+
+		lm := pubsub.NewLatencyMeasurer(logger)
+		ps, done := newPubsub()
+		defer done()
+
+		slow := newDelayedPublisher(ps, time.Second)
+		fast := newDelayedPublisher(ps, time.Nanosecond)
+		hold := make(chan struct{}, 1)
+
+		// Start two goroutines in which two subscribers are registered but the messages are received out-of-order because
+		// the first Pubsub will publish its message slowly and the second will publish it quickly. Both will ultimately
+		// receive their desired messages, but the slow publisher will receive an unexpected message first.
+		var wg sync.WaitGroup
+		wg.Add(2)
+		go func() {
+			ctx, cancel := context.WithTimeout(context.Background(), testutil.WaitShort)
+			defer cancel()
+			defer wg.Done()
+
+			hold <- struct{}{}
+			send, recv, err := lm.Measure(ctx, slow)
+			assert.NoError(t, err)
+			assert.Greater(t, send, 0.0)
+			assert.Greater(t, recv, 0.0)
+
+			// The slow subscriber will complete first and receive the fast publisher's message first.
+			logger.Sync()
+			assert.Contains(t, buf.String(), "received unexpected message")
+		}()
+
+		go func() {
+			ctx, cancel := context.WithTimeout(context.Background(), testutil.WaitShort)
+			defer cancel()
+			defer wg.Done()
+
+			// Force fast publisher to start after the slow one to avoid flakiness.
+			<-hold
+			time.Sleep(time.Millisecond * 50)
```

Comment:
YOLO-ed this a bit because Go's scheduler gives a goroutine at most 20ms on CPU before it is pre-empted; I added a bit of extra time to be sure. I ran this test with `-count=100` and didn't observe any flakes, but I'm open to better approaches.

Reply:
This feels too brittle to me.  Even if you ran it 100 times locally, we've observed really big variances in scheduling on CI.

I think you should simplify and _fake_ the PubSub, not wrap it with a delay.  Then you only need to have a single measurer, and you send it a bogus ping prior to sending the real one it published.  Added bonus is that is will execute in microseconds not a second.

Reply:
https://github.com/coder/coder/pull/13126/commits/633365db3a81498c861207a950dc22f43ab19ecb